benchmark: Add sleepBetweenRPCs and connections parameters

s-matyukevich opened the pull request:
I am trying to fix https://github.com/grpc/grpc-go/issues/5759 and have a branch with some promising results. However, the problem is that our current benchmarks don't really test the case I am trying to address. The case is the following: we have some grpc servers with a lot of mostly idle connections (in our case those connections mostly come from healtchecks) In this case grpc-go wastes a lot of memory on transport buffers, that aren't shared between connections. 

This PR modifies benchmarks so that we can simulate this case. It adds a random delay between subsequent RPCs and a configuration parameter which controls the maximum value of this delay. It also adds a parameter that allows configuring the number of connections. 

RELEASE NOTES: none

s-matyukevich commented on benchmark/benchmain/main.go:
```diff
@@ -244,6 +247,9 @@ func unconstrainedStreamBenchmark(start startFunc, stop ucStopFunc, bf stats.Fea
 				atomic.AddUint64(&req, 1)
+				if maxSleep > 0 {
+					time.Sleep(time.Duration(math_rand.Intn(maxSleep)))
+				}
```
Note that delay must be random. I first tried using a fixed delay and my optimization actually showed regression instead of improvement. The reason was that all concurrent goroutines send their requests exactly at the same time, and in such case the fact that we can share buffers between connections doesn't help as all connections are utilized simultaneously. After I modified the delay to be random I saw a big improvement in memory consumption.

s-matyukevich commented:
Here are the results of using the new option.

```
go run benchmark/benchmain/main.go -workloads all -kbps 0 -latency 0s -maxConcurrentCalls 1000 -respSizeBytes 1024 -reqSizeBytes 1024 -benchtime 1m -sleepBetweenRPCs 0s,100ms,5s 
go1.20.3/grpc1.56.0-dev
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 0s-:
50_Latency: 6837.4580µs 90_Latency: 8775.7080µs 99_Latency: 11084.3750µs        Avg_Latency: 7148.9370µs        Bytes/op: 17641.176256838167    Allocs/op: 156.56104943935168
Histogram (unit: µs)
Count: 8389127  Min: 205.8  Max: 32970.2  Avg: 7148.94
------------------------------------------------------------
[     205.792000,      205.793000)        1    0.0%    0.0%  
[     205.793000,      205.798840)        0    0.0%    0.0%  
[     205.798840,      205.838783)        0    0.0%    0.0%  
[     205.838783,      206.111988)        0    0.0%    0.0%  
[     206.111988,      207.980664)        0    0.0%    0.0%  
[     207.980664,      220.762072)        0    0.0%    0.0%  
[     220.762072,      308.184621)       21    0.0%    0.0%  
[     308.184621,      906.139262)      165    0.0%    0.0%  
[     906.139262,     4996.042354)    41167    0.5%    0.5%  
[    4996.042354,    32970.250000)  8347772   99.5%  100.0%  ##########
[   32970.250000,   224308.840623)        1    0.0%  100.0%  
Number of requests:  8389127    Request throughput:  1.1453954730666666e+09 bit/s
Number of responses: 8389127    Response throughput: 1.1453954730666666e+09 bit/s

go1.20.3/grpc1.56.0-dev
streaming-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 0s-:
50_Latency: 2343.4580µs 90_Latency: 5527.3330µs 99_Latency: 7059.8750µs Avg_Latency: 2829.4580µs        Bytes/op: 9624.39087511789      Allocs/op: 25.80684810536492
Histogram (unit: µs)
Count: 21153150  Min:  31.2  Max: 22563.3  Avg: 2829.46
------------------------------------------------------------
[      31.167000,       31.168000)         1    0.0%    0.0%  
[      31.168000,       31.173561)         0    0.0%    0.0%  
[      31.173561,       31.210048)         0    0.0%    0.0%  
[      31.210048,       31.449445)         0    0.0%    0.0%  
[      31.449445,       33.020158)         0    0.0%    0.0%  
[      33.020158,       43.325795)         1    0.0%    0.0%  
[      43.325795,      110.942333)         4    0.0%    0.0%  
[     110.942333,      554.582685)       636    0.0%    0.0%  
[     554.582685,     3465.361098)  17728124   83.8%   83.8%  ########
[    3465.361098,    22563.333000)   3424383   16.2%  100.0%  ##
[   22563.333000,   147867.454127)         1    0.0%  100.0%  
Number of requests:  21153150   Request throughput:  2.88811008e+09 bit/s
Number of responses: 21153150   Response throughput: 2.88811008e+09 bit/s

go1.20.3/grpc1.56.0-dev
unconstrained-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 0s-:
Bytes/op: 7326.442176775494     Allocs/op: 21.481295759919586
Histogram (empty)
Number of requests:  20433473   Request throughput:  2.789850180266667e+09 bit/s
Number of responses: 20273822   Response throughput: 2.7680524970666666e+09 bit/s

go1.20.3/grpc1.56.0-dev
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 100ms-:
50_Latency: 62.9580µs   90_Latency: 106.4170µs  99_Latency: 704.3750µs  Avg_Latency: 125.2880µs Bytes/op: 18180.5286051417      Allocs/op: 190.03853058496753
Histogram (unit: µs)
Count: 1197075  Min:  35.0  Max: 44531.1  Avg: 125.29
------------------------------------------------------------
[      35.041000,       35.042000)        1    0.0%    0.0%  
[      35.042000,       35.048076)        0    0.0%    0.0%  
[      35.048076,       35.091076)        0    0.0%    0.0%  
[      35.091076,       35.395357)        0    0.0%    0.0%  
[      35.395357,       37.548576)       19    0.0%    0.0%  
[      37.548576,       52.785659)   171808   14.4%   14.4%  #
[      52.785659,      160.609641)   950778   79.4%   93.8%  ########
[     160.609641,      923.617302)    67907    5.7%   99.5%  #
[     923.617302,     6322.979118)     5067    0.4%   99.9%  
[    6322.979118,    44531.125000)     1494    0.1%  100.0%  
[   44531.125000,   314907.969791)        1    0.0%  100.0%  
Number of requests:  1197075    Request throughput:  1.6344064e+08 bit/s
Number of responses: 1197075    Response throughput: 1.6344064e+08 bit/s

go1.20.3/grpc1.56.0-dev
streaming-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 100ms-:
50_Latency: 45.7920µs   90_Latency: 65.0000µs   99_Latency: 411.8330µs  Avg_Latency: 63.8210µs  Bytes/op: 10152.84408820177     Allocs/op: 59.083419696036664
Histogram (unit: µs)
Count: 1198434  Min:  22.7  Max: 8829.2  Avg: 63.82
------------------------------------------------------------
[     22.667000,      22.668000)        1    0.0%    0.0%  
[     22.668000,      22.672911)        0    0.0%    0.0%  
[     22.672911,      22.701937)        0    0.0%    0.0%  
[     22.701937,      22.873507)        0    0.0%    0.0%  
[     22.873507,      23.887622)       12    0.0%    0.0%  
[     23.887622,      29.881834)     6934    0.6%    0.6%  
[     29.881834,      65.312344)  1073836   89.6%   90.2%  #########
[     65.312344,     274.734522)    97051    8.1%   98.3%  #
[    274.734522,    1512.584298)    19031    1.6%   99.9%  
[   1512.584298,    8829.250000)     1568    0.1%  100.0%  
[   8829.250000,   52076.498613)        1    0.0%  100.0%  
Number of requests:  1198434    Request throughput:  1.636261888e+08 bit/s
Number of responses: 1198434    Response throughput: 1.636261888e+08 bit/s

go1.20.3/grpc1.56.0-dev
unconstrained-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 100ms-:
Bytes/op: 16027.039590154656    Allocs/op: 67.7413706509481
Histogram (empty)
Number of requests:  1192546    Request throughput:  1.6282228053333333e+08 bit/s
Number of responses: 1192744    Response throughput: 1.6284931413333333e+08 bit/s

go1.20.3/grpc1.56.0-dev
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 5s-:
50_Latency: 416.0000µs  90_Latency: 1006.0000µs 99_Latency: 8585.0410µs Avg_Latency: 819.9940µs Bytes/op: 18358.007899987784    Allocs/op: 201.75986480433278
Histogram (unit: µs)
Count: 24557  Min:  78.0  Max: 15476.4  Avg: 819.99
------------------------------------------------------------
[      78.000000,       78.001000)      1    0.0%    0.0%  
[      78.001000,       78.006289)      0    0.0%    0.0%  
[      78.006289,       78.039556)      0    0.0%    0.0%  
[      78.039556,       78.248786)      0    0.0%    0.0%  
[      78.248786,       79.564710)      0    0.0%    0.0%  
[      79.564710,       87.841069)      1    0.0%    0.0%  
[      87.841069,      139.894309)     41    0.2%    0.2%  
[     139.894309,      467.277382)  14602   59.5%   59.6%  ######
[     467.277382,     2526.316861)   8625   35.1%   94.8%  ####
[    2526.316861,    15476.417000)   1286    5.2%  100.0%  #
[   15476.417000,    96924.633645)      1    0.0%  100.0%  
Number of requests:  24557      Request throughput:  3.352849066666667e+06 bit/s
Number of responses: 24557      Response throughput: 3.352849066666667e+06 bit/s

go1.20.3/grpc1.56.0-dev
streaming-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 5s-:
50_Latency: 313.8330µs  90_Latency: 692.5410µs  99_Latency: 5077.1250µs Avg_Latency: 523.4040µs Bytes/op: 10382.226754066522    Allocs/op: 70.62713441773893
Histogram (unit: µs)
Count: 24714  Min:  38.9  Max: 17024.2  Avg: 523.40
------------------------------------------------------------
[      38.916000,       38.917000)      1    0.0%    0.0%  
[      38.917000,       38.922358)      0    0.0%    0.0%  
[      38.922358,       38.956428)      0    0.0%    0.0%  
[      38.956428,       39.173054)      0    0.0%    0.0%  
[      39.173054,       40.550430)      0    0.0%    0.0%  
[      40.550430,       49.308208)     14    0.1%    0.1%  
[      49.308208,      104.992860)    243    1.0%    1.0%  
[     104.992860,      459.053029)  18558   75.1%   76.1%  ########
[     459.053029,     2710.276657)   5019   20.3%   96.4%  ##
[    2710.276657,    17024.250000)    878    3.6%  100.0%  
[   17024.250000,   108036.902084)      1    0.0%  100.0%  
Number of requests:  24714      Request throughput:  3.3742848e+06 bit/s
Number of responses: 24714      Response throughput: 3.3742848e+06 bit/s

go1.20.3/grpc1.56.0-dev
unconstrained-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1000-reqSize_1024B-respSize_1024B-compressor_off-channelz_false-preloader_false-clientReadBufferSize_-1-clientWriteBufferSize_-1-serverReadBufferSize_-1-serverWriteBufferSize_-1-sleepBetweenRPCs 5s-:
Bytes/op: 685789.845895303      Allocs/op: 2470.9679858835393
Histogram (empty)
Number of requests:  23854      Request throughput:  3.2568661333333333e+06 bit/s
Number of responses: 23750      Response throughput: 3.2426666666666665e+06 bit/s
```

dfawley commented on benchmark/benchmain/main.go:
```diff
@@ -255,6 +261,9 @@ func unconstrainedStreamBenchmark(start startFunc, stop ucStopFunc, bf stats.Fea
 				}
 				recver(pos)
 				atomic.AddUint64(&resp, 1)
+				if maxSleep > 0 {
+					time.Sleep(time.Duration(math_rand.Intn(maxSleep)))
+				}
```
Is this client code or server code, or is this used for both sides?

I think we might want the sender in both directions to be sleeping, and not sleeping after receiving.  If the sender isn't also sleeping, it could fill up buffers and have more impact than you intended.

s-matyukevich replied:
Yes, you are right. I'll remove the delay from the client receive goroutine and instead will add it [here](https://github.com/grpc/grpc-go/blob/098b2d00c5bcbc5c696dcab9a2a7f7b442f8142e/benchmark/benchmark.go#L137-L145)

In order to do that I'll need to pass the sleepBetweenRPCs parameter to the server (probably I'll add ti to SimpleRequest rpc)

s-matyukevich replied:
Fixed. I end up using metadata to pass the sleep duration to the server. [SimpleRequest](https://github.com/grpc/grpc-proto/blob/master/grpc/testing/messages.proto#L69) protobuf is used in too many places across languages, so I am not sure if adding fields to it is a good idea.

dfawley reviewed (commented):
LGTM, but please see the question/comment below.  Thanks!

s-matyukevich commented:
@dfawley  I fixed your comment, but I also discovered 2 more issues:

1. Right now benchmarks create a single connection between client and server, which makes it impossible to reproduce the case I  described in the PR description. I added one more parameter which allows to use a separate connection per goroutine. Without this change my original PR is completely useless, but let me know if I should split those 2 changes into separate PRs.
2. benchmarks memory reporting is actually not very useful. We compare total alocations before and after benchmark run and divide them by the number of RPC. This includes allocation that are actually garbage collected and this doesn't give you an idea how much memory your benchamrk is actually using. IMO in order to measure the actual memory usage we need to sample [MemStats.Alloc](https://go.dev/src/runtime/mstats.go) during program execution with some relatively high frequency and then take the average (which is something that pprof does) For now I am just using pprof to analyze memory usage, but let me know if we should add some useful memory-usage indicator to the benchmark output.

s-matyukevich commented on benchmark/benchmain/main.go:
```diff
@@ -346,16 +353,41 @@ func makeClient(bf stats.Features) (testgrpc.BenchmarkServiceClient, func()) {
+		tc = clients[0]
+	}
+	return tc
+}
```
We can slightly simplify the code and remove this function if we always create `MaxConcurrentCalls` clients, but in case when `ShareConnection == true` all those clients will use a single connection. I didn't implement this options because it modifies the behavior of benchmarks with default setting, which I want to avoid.

dfawley commented on benchmark/benchmain/main.go:
```diff
@@ -111,6 +111,7 @@ var (
 	serverReadBufferSize  = flags.IntSlice("serverReadBufferSize", []int{-1}, "Configures the server read buffer size in bytes. If negative, use the default - may be a a comma-separated list")
 	serverWriteBufferSize = flags.IntSlice("serverWriteBufferSize", []int{-1}, "Configures the server write buffer size in bytes. If negative, use the default - may be a a comma-separated list")
 	sleepBetweenRPCs      = flags.DurationSlice("sleepBetweenRPCs", []time.Duration{0}, "Configures the maximum amount of time the client should sleep between consecutive RPCs - may be a a comma-separated list")
+	shareConnection       = flag.Bool("shareConnection", true, "Controls whether a single connection or connection per `maxConcurrentCalls` should be used.")
```
Instead of this, how about a flag for the number of connections, where each connection does `maxConcurrentCalls` at a time?  Hopefully that wouldn't be too different implementation-wise, and would allow for some more interesting testing options.

s-matyukevich replied:
Sure, done.

s-matyukevich commented on benchmark/benchmark.go:
```diff
@@ -213,7 +233,7 @@ func StartServer(info ServerInfo, opts ...grpc.ServerOption) func() {
 	go s.Serve(info.Listener)
 	return func() {
-		s.Stop()
+		s.GracefulStop()
```
This is an unrelated fix. I run a lot of benchmarks recently while testing https://github.com/grpc/grpc-go/pull/6309

The error that I see very often is `2023/05/24 15:40:52 server send error: rpc error: code = Canceled desc = context canceled`, which is very annoying. This happens on my branch as well as on master. 

This was an attempt to fix this error, but it doesn't help. I can rollback this line if you prefer.

dfawley replied:
Yes, I think it would make more sense to leave this as `Stop` since `GracefulStop` can technically leave the server running forever if there is an open stream.

We should be able to get to the bottom of the context canceled errors, but I don't remember seeing it before -- is it happening in the stdout logging or elsewhere? Does it cause any problems when it happens, e.g. prevents output files from being written?

s-matyukevich replied:
I reverted this change. 

The error happens [here](https://github.com/grpc/grpc-go/blob/master/benchmark/benchmark.go#L131) and the app panics without reporting any results.

I also sometimes see one more type of errors `2023/05/24 16:38:05 FATAL: [benchmark] DoUnaryCall failed: /BenchmarkService/UnaryCall(_, _) = _, rpc error: code = Unavailable desc = connection error: desc = "error reading server preface: read tcp 127.0.0.1:48722->127.0.0.1:43533: use of closed network connection", want _, <nil>`  which also stops the benchmarks. I didn't see that one on master.

s-matyukevich replied:
I fixed it here https://github.com/grpc/grpc-go/pull/6299/files#r1204793848

s-matyukevich commented on benchmark/benchmark.go:
```diff
@@ -135,9 +152,12 @@ func (s *testServer) UnconstrainedStreamingCall(stream testgrpc.BenchmarkService
 			err := stream.Send(response)
 			switch status.Code(err) {
-			case codes.Unavailable:
+			case codes.Unavailable, codes.Canceled:
```
I think there is a race here. If server is in the process of sending response while the client closes the connection, we will get Canceled, instead of Unavailable.  Unavailable is still a lot more common. 

This error can be reproduced on master, but with my PR it became a lot more apparent, because now you can increase the number of connections, which greatly increase the chance that at least one connection will result in Canceled.

s-matyukevich commented:
> This PR is labeled as requiring an update from the reporter, and no update has been received after 6 days. If no update is provided in the next 7 days, this issue will be automatically closed.

I think I answered all questions and implemented all suggestions, what other updates are expected?

dfawley reviewed (approved):
Thank you!

dfawley commented:
We need a second review before merging PRs from external contributors.

zasweq commented on benchmark/stats/stats.go:
```diff
@@ -121,6 +124,8 @@ type Features struct {
 	ServerReadBufferSize int
 	// ServerWriteBufferSize is the size of the server write buffer in bytes. If negative, use the default buffer size.
 	ServerWriteBufferSize int
+	// SleepBetweenRPCs configures optional delay between RPCs
```
Nit: period at end.

zasweq commented on benchmark/benchmain/main.go:
```diff
@@ -461,32 +491,45 @@ func streamCaller(stream testgrpc.BenchmarkService_StreamingCallClient, reqSize,
-	// Warm up connection.
-	for i := 0; i < warmupCallCount; i++ {
-		caller(0)
+	// if SleepBetweenRPCs > 0 we skip the warmup because otherwise
+	// we are going to send a set of simultaneous requests on every connection,
+	// which is something we are trying to avoid when using SleepBetweenRPCs param.
```
Nit: wrap comments to 80 characters please.

s-matyukevich replied:
done.

zasweq commented on benchmark/benchmark.go:
```diff
@@ -27,6 +27,9 @@ import (
 	"io"
 	"log"
 	"net"
+	"time"
+
+	"math/rand"
```
Please group these imports together. You also don't rename this import, which is what I think we should do for main.go as well.

s-matyukevich replied:
done

zasweq commented on benchmark/benchmain/main.go:
```diff
@@ -47,6 +47,7 @@ import (
 	"fmt"
 	"io"
 	"log"
+	math_rand "math/rand"
```
Please just leave this import name as is.

s-matyukevich replied:
done

zasweq commented on benchmark/benchmain/main.go:
```diff
@@ -389,51 +411,59 @@ func makeFuncStream(bf stats.Features) (rpcCallFunc, rpcCleanupFunc) {
-			streams[pos].Send(req)
-		}, func(pos int) {
-			streams[pos].Recv()
+	return func(conn, pos int) {
```
Nit: I see conn and cn throughout this PR for the first dimension of conn index (whether a function parameter or an index declaration in a for iteration). I also see pos and i for the steam index within the conn. Please make consistent across PR.

s-matyukevich replied:
Done, used cn and pos thought the PR.

zasweq reviewed (approved):
Some minor nits. LGTM otherwise though. Please fix before merging.

zasweq reviewed (approved):
Some minor nits. LGTM otherwise though. Please get to those before merging :). Test failure is a flake, not related to this PR.

zasweq commented:
Awesome, thanks for getting to my comments. Feel free to merge (which I think unblocks your other issue) :).

s-matyukevich commented:
> Awesome, thanks for getting to my comments. Feel free to merge (which I think unblocks your other issue) :).

I don't have permissions to merge - please do it for me.

zasweq commented:
Done :).
